source/Irrlicht/CIrrDeviceSDL.cpp

```diff
@@ -908,6 +933,83 @@ void CIrrDeviceSDL::setWindowCaption(const wchar_t* text)
 }
 
 
+//! Sets the window icon.
+bool CIrrDeviceSDL::setWindowIcon(const std::string &path)
+{
+	if (!Window)
+		return false;
+
+	video::IImageLoader *image_loader = nullptr;
+	u32 cnt = VideoDriver->getImageLoaderCount();
+	for (u32 i = 0; i < cnt; i++) {
+		if (VideoDriver->getImageLoader(i)->isALoadableFileExtension(
+					path.c_str())) {
+			image_loader = VideoDriver->getImageLoader(i);
+			break;
+		}
+	}
+
+	if (!image_loader) {
+		os::Printer::log(("Could not find image loader for file '" + path + "'").c_str(), ELL_WARNING);
+		return false;
+	}
+
+	io::IReadFile *icon_f = getFileSystem()->createAndOpenFile(path.c_str());
+
+	if (!icon_f) {
+		os::Printer::log(("Could not load icon file '" + path + "'").c_str(), ELL_WARNING);
+		return false;
+	}
+
+	video::IImage *img = image_loader->loadImage(icon_f);
+
+	if (!img) {
+		os::Printer::log(("Could not load icon file '" + path + "'").c_str(), ELL_WARNING);
+		icon_f->drop();
+		return false;
+	}
+
+	u32 height = img->getDimension().Height;
+	u32 width = img->getDimension().Width;
+
+	SDL_Surface *surface = SDL_CreateRGBSurface(0, width, height, 32,
+			0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000);
+
+	if (!surface) {
+		os::Printer::log("Failed to create SDL suface", ELL_ERROR);
+		img->drop();
+		icon_f->drop();
+		return false;
+	}
+
+	SDL_LockSurface(surface);
+
+	for (u32 y = 0; y < height; y++) {
+		u32 *row = (u32 *)((u8 *)surface->pixels + y * surface->pitch);
+		for (u32 x = 0; x < width; x++) {
+			video::SColor col = img->getPixel(x, y);
+			u32 pixel_val = 0;
+			pixel_val |= (u32)col.getAlpha() << 24;
+			pixel_val |= (u32)col.getRed() << 16;
+			pixel_val |= (u32)col.getGreen() << 8;
+			pixel_val |= (u32)col.getBlue();
+			row[x] = pixel_val;
+		}
+	}

```

Comment:
Can’t you convert the image to ARGB with e.g. `IImage::copyTo`? Or at least use `col.color`, it is in the same format as `pixel_val` IIUC.

Reply:
Same as above.

(But we still have to manually copy for X11, because it's an array of longs.)